test/replayTools.js

```diff
@@ -0,0 +1,74 @@
+'use strict';
+module.exports = {
+  toSaver: wrapHelperToSave
+};
+
+var fs = require('fs');
+var path = require('path');
+
+/**
+ * Monkey-patch a Helper constructor to spy
+ * on data from Algolia and save the server response.
+ * This is used to created data for tests.
+ * @param {Helper} Helper a vanilla Helper constructor function
+ * @return {HelperSaver} a Helper enhnaced with save function
+ */
+function wrapHelperToSave(Helper, folderName) {
+  var savedParameters = {};
+
+  createFolderIfNone(folderName);
+
+  function WrappedHelper() {
+    Helper.apply(this, Array.prototype.slice.call(arguments));
+  }
+
+  WrappedHelper.prototype = Object.create(Helper.prototype);
+  WrappedHelper.prototype.constructor = WrappedHelper;
+  WrappedHelper.prototype._handleResponse = function(state, queryId, err, content) {
+    savedParameters.state = state;
+    savedParameters.error = err;
+    savedParameters.content = content;
+
+    Helper.prototype._handleResponse.apply(this, Array.prototype.slice.call(arguments));
+  };
+  WrappedHelper.prototype.searchOnce = function() {
+    return Helper.prototype.searchOnce.apply(
+      this,
+      Array.prototype.slice.call(arguments)
+    ).then(function(contentAndState) {
+      savedParameters.content = contentAndState._originalResponse;
+      savedParameters.state = contentAndState.state;
+      savedParameters.error = null;
+      return contentAndState;
+    }, function(error) {
+      savedParameters.content = null;
+      savedParameters.state = null; // FIXME can't access the actual state without reimplementation
```

Comment:
You could maybe compute the state beforehand then pass it as the first parameter to searchOnce ?